---
title: "Dotplot with inferCNV example dataset"
author: "Simon Leonard"
date: "22/02/01"
output:
  html_document:
    toc: yes
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{FlexDotPlot_InferCNV_dataset}
  %\VignetteEngine{knitr::knitr}
  %\VignetteEncoding{UTF-8}
---

## Package Loading
```{r message=FALSE}
library(infercnv)
library(reshape2)
library(biomaRt)
library(Seurat)
library(ggplot2)
```

## Compute InferCNV data

### Run InferCNV with example dataset 

We run InferCNV as described in the original tutorial (https://github.com/broadinstitute/inferCNV/wiki)

```{r eval=FALSE, message=FALSE, warning=FALSE}

infercnv_obj = CreateInfercnvObject(raw_counts_matrix=system.file("extdata", "oligodendroglioma_expression_downsampled.counts.matrix.gz", package = "infercnv"),
                                    annotations_file = system.file("extdata", "oligodendroglioma_annotations_downsampled.txt", package = "infercnv"), 
                                    delim="\t", 
                                    gene_order_file = system.file("extdata", "gencode_downsampled.EXAMPLE_ONLY_DONT_REUSE.txt", package = "infercnv"),
                                    ref_group_names = c("Microglia/Macrophage","Oligodendrocytes (non-malignant)")) 


infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=1, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics
                             out_dir="output", 
                             cluster_by_groups=TRUE, 
                             denoise=TRUE,
                             HMM=TRUE)
```


### Process InferCNV output 

#### Import InferCNV score output

```{r}
e=read.table("E:/Scripts_DEPRECATED/Infer_cnv_example_dataset/output_subclusters/infercnv.observations.txt", header=T, row.names = 1)

include_ref=T
if (include_ref){
  ref=read.table("E:/Scripts_DEPRECATED/Infer_cnv_example_dataset/output_subclusters/infercnv.references.txt", header=T, row.names = 1)
  e=cbind(e,ref)
}

e2=setNames(melt(as.matrix(e)), c("gene","cell","modified_expr"))
e2$cell=gsub("\\.","-",e2$cell)

e2$centered_expr=e2$modified_expr
# hist(e2$centered_expr, breaks=50)
```


#### Import chromosom band coordinates

Link to download the table : http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/cytoBand.txt.gz

```{r}
bands=read.table("E:/chrom_band_positions_hg38.txt", sep="\t", header=F)
colnames(bands)=c("chrom","start","end","band","type")
bands=bands[!grepl("_",bands$chrom),]
bands$chrom=gsub("chr","",bands$chrom)

arm_resolution=T
if (arm_resolution){
  bands$b1=paste(bands$chrom, substring(bands$band,1,1), sep="_")
  r1=aggregate(bands$start, list(bands$b1), min)
  r2=aggregate(bands$end, list(bands$b1), max)
  bands=data.frame(colsplit(r1[,1],"_",c("chrom","band")),start=r1[,2], end=r2[,2])
}

```



#### Get genes coordinates + assign each gene to an arm

```{r}
ensembl = biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")
gene_positions <- biomaRt::getBM(attributes = c("ensembl_gene_id", 
                                                "hgnc_symbol", "chromosome_name", "start_position", 
                                                "end_position"), filters = "hgnc_symbol", values = unique(e2$gene), 
                                 mart = ensembl)
gene_positions=gene_positions[!grepl("_",gene_positions$chromosome_name),]

gene_positions$band=NA
for (i in 1:nrow(gene_positions)){
  count=sum(gene_positions[i,3]==bands$chrom & gene_positions[i,4]>=bands$start & gene_positions[i,5]<=bands$end)
  if(count==1){
    count=which(gene_positions[i,3]==bands$chrom & gene_positions[i,4]>=bands$start & gene_positions[i,5]<=bands$end)
    gene_positions$band[i]=as.character(bands$band[count])
  }else if (count==0){
    gene_positions$band[i]=("no_band")
  }else {
    gene_positions$band[i]=("multiple_band")
  }
}

gene_positions$chrom_band=paste(gene_positions$chromosome_name, gene_positions$band, sep="_")
gene_positions2=gene_positions[gene_positions$band!="no_band",]
```


#### Aggregate InferCNV scores to have one mean per cell per arm


```{r}
e2$chrom_band=gene_positions2$chrom_band[match(e2$gene, gene_positions2$hgnc_symbol)]

# Output = 3 columns dataframe (cell, band, value)
e3=aggregate(e2$centered_expr, e2[,c("cell","chrom_band")], FUN=mean)
colnames(e3)[3]="mean_centered_modified_expr"

# Output matrix cell x band
e4=acast(e3, cell~chrom_band, value.var="mean_centered_modified_expr")
```


## Get corresponding processed scRNAseq data 

Link to download the dataset : https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70630


```{r}
expr=read.table("E:/Scripts_DEPRECATED/Infer_cnv_example_dataset/GSE70630_OG_processed_data_v2.txt", header=T, row.names = 1)

#restric to cells and gene from example infercnv dataset
expr=expr[rownames(e),colnames(e)]
```


## Generate DotPlot

Create a gene list to perform DotPlot

```{r}
genes=c("PCDHGB2", "PCDHGB4","HSPA1A","ARRDC2","NEAT1","GPM6A","UGDH-AS1","CCND1","GRIA2","SCD5",
        "MTRNR2L6","PDGFRA","PRDX1",
        "EEF1G","LINC00689","SLC35F6","PCDH9","PTN","ARPC3","PFDN5","ITM2B",
        "SPP1","PTPRZ1","BCAN","FXYD6","KLRC2","SLC43A3","TIMM10B",
        "TPT1","UBL3","FNDC3A")

```

Get DotPlot data with Seurat DotPlot function

```{r}
seurat=CreateSeuratObject(expr)

seurat$sample=sapply(Cells(seurat), function(x){gsub("^X","",substring(x, 1, as.numeric(gregexpr("_",x)[[1]])[1]-1))})
seurat$type=seurat$sample
seurat$type[colnames(infercnv_obj@expr.data)[infercnv_obj@reference_grouped_cell_indices$`Microglia/Macrophage`]]="Microglia/Macrophage"
seurat$type[colnames(infercnv_obj@expr.data)[infercnv_obj@reference_grouped_cell_indices$`Oligodendrocytes (non-malignant)`]]="Oligodendrocytes (non malignant)"

seurat$malignant=ifelse(seurat$type %in% c("Microglia/Macrophage","Oligodendrocytes (non malignant)"), "non malignant","malignant")
DotPlot(seurat, features = genes, group.by = "type")+scale_colour_gradientn(colours=c("blue","white","red", "darkred"))+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
d=DotPlot(seurat, features = genes, group.by = "type")$data
d=cbind(d[,c(4,3)], d[,c(1,2,5)])
```

Add gene bands and cnv scores

```{r}
#add gene bands
d$band=gene_positions2$chrom_band[match(d$features.plot, gene_positions2$hgnc_symbol)]

# add cnv scores
e3$type=seurat$type[match(e3$cell, rownames(seurat@meta.data))]
means=aggregate(e3$mean_centered_modified_expr, by=list(e3$chrom_band, e3$type), FUN=mean)
colnames(means)=c("band","type","mean")

d$mean_band_cnv_score=sapply(1:nrow(d), function(x){
  means$mean[means$band==d$band[x] & means$type==d$id[x]]
})

# Compute cnv status according to cnv scores
d$cnv_status=ifelse(d$mean_band_cnv_score<0.95, 
                    "down", ifelse(d$mean_band_cnv_score>1.05,
                                   "up","normal"))
d$cnv_status=factor(d$cnv_status, levels = c("up","normal","down"))
```

Perform DotPlot

```{r message=FALSE, warning=FALSE, fig.width= 14, fig.height=6}
library(FlexDotPlot)
dotplot=dot_plot(d, size_var = "pct.exp", size.breaks.number = 5,
         col_var = "avg.exp.scaled", cols.use = c("blue","white","red", "darkred"), color.breaks.values = c(-1,0,1),
         shape_var = "cnv_status", shape_use = c( "\u2BC5", "\u25CF", "\u2BC6"), 
         x.lab.pos = "bottom", transpose = T, y.lab.size.factor = 0.3,
         dend_x_var = c("avg.exp.scaled","pct.exp"), dend_y_var = c("avg.exp.scaled","pct.exp"), do.return = T)
```
## Dotplot with Seurat (without cnv status) and same gene/sample ordering
```{r}
seurat$type=factor(seurat$type, levels=dotplot$plot$grobs[[4]]$label)
DotPlot(seurat, features = dotplot$plot$grobs[[10]]$label, group.by = "type")+scale_colour_gradientn(colours=c("blue","white","red", "darkred"))+theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

## Session Info
```{r}
sessionInfo()
```
